executor/src/witgen/jit/witgen_inference.rs

```diff
@@ -0,0 +1,368 @@
+
+    /// Process an identity on a certain row.
+    /// Returns true if this identity/row pair was fully processed and
+    /// should not be considered again.
+    pub fn process_identity(&mut self, id: &Identity<T>, row_offset: i32) -> bool {
+        let result = match id {
+            Identity::Polynomial(PolynomialIdentity { expression, .. }) => {
+                self.process_polynomial_identity(expression, row_offset)
+            }
+            Identity::Lookup(LookupIdentity {
+                id,
+                source: _,
+                left,
+                right,
+            })
+            | Identity::PhantomLookup(PhantomLookupIdentity {
+                id,
+                source: _,
+                left,
+                right,
+                multiplicity: _,
+            }) => {
+                // TODO multiplicity?
+                self.process_lookup(*id, left, right, row_offset)
+            }
+            _ => {
+                // TODO
+                ProcessResult::empty()
+            }
+        };
+        self.ingest_effects(result.effects);
+        result.complete
+    }
+
+    fn process_polynomial_identity(
+        &self,
+        expression: &'a Expression<T>,
+        offset: i32,
+    ) -> ProcessResult<T, Cell> {
+        if let Some(r) = self.evaluate(expression, offset) {
+            // TODO propagate or report error properly.
+            // If solve returns an error, it means that the constraint is conflicting.
+            // In the future, we might run this in a runtime-conditional, so an error
+            // could just mean that this case cannot happen in practice.
+            // TODO this can also happen if we solve a constraint where all values are
+            // known but symbolic.
+            r.solve(self).unwrap()
+        } else {
+            ProcessResult::empty()
+        }
+    }
+
+    fn process_lookup(
+        &self,
+        lookup_id: u64,
+        left: &SelectedExpressions<T>,
+        right: &SelectedExpressions<T>,
+        offset: i32,
+    ) -> ProcessResult<T, Cell> {
+        // TODO: In the future, call the 'mutable state' to check if the
+        // lookup can always be answered.
+
+        // If the RHS is fully fixed columns...
+        if right.expressions.iter().all(|e| match e {
+            Expression::Reference(r) => r.is_fixed(),
+            Expression::Number(_) => true,
+            _ => false,
+        }) {
+            // and the selector is known to be 1...
+            if self
+                .evaluate(&left.selector, offset)
+                .and_then(|s| s.try_to_known().map(|k| k.is_known_one()))
+                == Some(true)
+            {
+                if let Some(lhs) = left
+                    .expressions
+                    .iter()
+                    .map(|e| self.evaluate(e, offset))
+                    .collect::<Option<Vec<_>>>()
+                {
+                    // and all except one expression is known on the LHS.
```

Comment:
Why this? I think there could be any number of outputs (including 0, e.g. for `PoseidonGLMemory`).

Reply:
Yes, this is a placeholder, see the TODO in line 101: ` // TODO: In the future, call the 'mutable state' to check if the
        // lookup can always be answered.`